File: MacOS/Type4Me/Audio/AudioCaptureEngine.swift
@preconcurrency import AVFoundation

enum AudioCaptureError: Error, LocalizedError {
    case converterCreationFailed
    case microphonePermissionDenied
    case noInputDevice

    var errorDescription: String? {
        switch self {
        case .converterCreationFailed:
            return L("录音启动失败", "Failed to start recording")
        case .microphonePermissionDenied:
            return L("未授予麦克风权限", "Microphone permission not granted")
        case .noInputDevice:
            return L("找不到麦克风", "No microphone found")
        }
    }
}

final class AudioCaptureEngine: NSObject, @unchecked Sendable, AVCaptureAudioDataOutputSampleBufferDelegate {

    // MARK: - Static properties

    static let sampleRate: Double = 16000
    static let channels: AVAudioChannelCount = 1
    static let chunkDurationMs: Int = 200
    static let samplesPerChunk: Int = 3200
    static let chunkByteSize: Int = 6400
    static let targetFormat: AVAudioFormat = AVAudioFormat(
        commonFormat: .pcmFormatInt16,
        sampleRate: 16000,
        channels: 1,
        interleaved: true
    )!

    // MARK: - Public

    var onAudioChunk: ((Data) -> Void)?
    var onAudioLevel: ((Float) -> Void)?

    // MARK: - Private

    private var captureSession: AVCaptureSession?
    private let bufferLock = NSLock()
    private var buffer = Data()
    private var accumulatedAudio = Data()
    private var converter: AVAudioConverter?
    private let outputQueue = DispatchQueue(label: "com.type4me.audiocapture")
    private var levelCounter = 0

    // MARK: - Warm-up

    private var isWarmedUp = false
    private var warmSession: AVCaptureSession?

    /// Pre-initialize the audio capture pipeline so the first real recording starts instantly.
    func warmUp() {
        guard !isWarmedUp else { return }
        outputQueue.async { [weak self] in
            guard let self else { return }
            do {
                guard let device = AVCaptureDevice.default(for: .audio) else { return }
                let session = AVCaptureSession()
                let input = try AVCaptureDeviceInput(device: device)
                guard session.canAddInput(input) else { return }
                session.addInput(input)
                let output = AVCaptureAudioDataOutput()
                guard session.canAddOutput(output) else { return }
                session.addOutput(output)
                session.startRunning()
                // Keep it alive briefly to fully initialize CoreAudio, then stop
                Thread.sleep(forTimeInterval: 0.3)
                session.stopRunning()
                self.isWarmedUp = true
                NSLog("[Audio] Warm-up complete")
            } catch {
                NSLog("[Audio] Warm-up failed: %@", String(describing: error))
            }
        }
    }

    // MARK: - Start / Stop

    func start() throws {
        let authStatus = AVCaptureDevice.authorizationStatus(for: .audio)
        guard authStatus == .authorized else {
            throw AudioCaptureError.microphonePermissionDenied
        }

        // Reset state
        bufferLock.lock()
        buffer = Data()
        accumulatedAudio = Data()
        bufferLock.unlock()
        converter = nil

        try startWithAVCapture()
    }

    private func startWithAVCapture() throws {
        let session = AVCaptureSession()

        guard let device = AVCaptureDevice.default(for: .audio) else {
            throw AudioCaptureError.noInputDevice
        }

        let input = try AVCaptureDeviceInput(device: device)
        guard session.canAddInput(input) else {
            throw AudioCaptureError.converterCreationFailed
        }
        session.addInput(input)

        let output = AVCaptureAudioDataOutput()
        output.setSampleBufferDelegate(self, queue: outputQueue)
        guard session.canAddOutput(output) else {
            throw AudioCaptureError.converterCreationFailed
        }
        session.addOutput(output)

        session.startRunning()
        captureSession = session
        isWarmedUp = true
        NSLog("[Audio] Capture session started (AVCapture), device: %@", device.localizedName)
    }

    func stop() {
        captureSession?.stopRunning()
        captureSession = nil
        converter = nil
        levelCounter = 0
        flushRemaining()
        NSLog("[Audio] Capture session stopped")
    }

    // MARK: - AVCaptureAudioDataOutputSampleBufferDelegate

    func captureOutput(
        _ output: AVCaptureOutput,
        didOutput sampleBuffer: CMSampleBuffer,
        from connection: AVCaptureConnection
    ) {
        guard let pcmBuffer = sampleBuffer.toPCMBuffer() else { return }

        // Emit audio level ~20 times/sec (every 3rd callback at typical 60Hz buffer rate)
        levelCounter += 1
        if levelCounter % 3 == 0, let onAudioLevel {
            let level = Self.calculateLevel(from: pcmBuffer)
            onAudioLevel(level)
        }

        // Lazy-create converter from source format → 16kHz Int16
        if converter == nil {
            let sourceFormat = pcmBuffer.format
            converter = AVAudioConverter(from: sourceFormat, to: Self.targetFormat)
            NSLog("[Audio] Input format: %@", sourceFormat.description)
        }

        guard let converter else { return }
        convert(buffer: pcmBuffer, using: converter)
    }

    // MARK: - Internal

    private func convert(buffer pcmBuffer: AVAudioPCMBuffer, using converter: AVAudioConverter) {
        let frameCapacity = AVAudioFrameCount(
            Double(pcmBuffer.frameLength) * Self.sampleRate / pcmBuffer.format.sampleRate
        )
        guard frameCapacity > 0 else { return }
        guard let convertedBuffer = AVAudioPCMBuffer(
            pcmFormat: Self.targetFormat,
            frameCapacity: frameCapacity
        ) else { return }

        var error: NSError?
        nonisolated(unsafe) var hasData = true
        let status = converter.convert(to: convertedBuffer, error: &error) { _, outStatus in
            if hasData {
                hasData = false
                outStatus.pointee = .haveData
                return pcmBuffer
            }
            outStatus.pointee = .noDataNow
            return nil
        }

        guard status != .error, error == nil else { return }

        let byteCount = Int(convertedBuffer.frameLength) * MemoryLayout<Int16>.size
        guard byteCount > 0 else { return }

        let audioBuffer = convertedBuffer.audioBufferList.pointee.mBuffers
        guard let mData = audioBuffer.mData else { return }
        let chunk = Data(bytes: mData, count: byteCount)

        bufferLock.lock()
        accumulatedAudio.append(chunk)
        buffer.append(chunk)
        emitFullChunks()
        bufferLock.unlock()
    }

    /// Returns the full recorded PCM audio since the last start().
    func getRecordedAudio() -> Data {
        bufferLock.lock()
        let data = accumulatedAudio
        bufferLock.unlock()
        return data
    }

    /// Emit all complete chunks from the buffer. Must be called with bufferLock held.
    private func emitFullChunks() {
        while buffer.count >= Self.chunkByteSize {
            let chunk = buffer.prefix(Self.chunkByteSize)
            buffer.removeFirst(Self.chunkByteSize)
            onAudioChunk?(Data(chunk))
        }
    }

    /// RMS → normalized 0..1 level from float PCM buffer.
    private static func calculateLevel(from buffer: AVAudioPCMBuffer) -> Float {
        guard let channelData = buffer.floatChannelData else { return 0 }
        let frames = Int(buffer.frameLength)
        guard frames > 0 else { return 0 }

        let ptr = channelData[0]
        var sum: Float = 0
        // Sample every 16th frame for efficiency (256 samples max)
        let stride = max(1, frames / 256)
        var count = 0
        var i = 0
        while i < frames {
            sum += ptr[i] * ptr[i]
            count += 1
            i += stride
        }
        let rms = sqrt(sum / Float(count))
        let db = 20 * log10(max(rms, 1e-7))
        // Map -50dB..0dB → 0..1
        return max(0, min(1, (db + 50) / 50))
    }

    private func bufferSize() -> Int {
        bufferLock.lock()
        let size = buffer.count
        bufferLock.unlock()
        return size
    }

    private func flushRemaining() {
        bufferLock.lock()
        let remaining = buffer
        buffer = Data()
        bufferLock.unlock()

        if !remaining.isEmpty {
            onAudioChunk?(remaining)
        }
    }
}

// MARK: - CMSampleBuffer → AVAudioPCMBuffer

private extension CMSampleBuffer {
    func toPCMBuffer() -> AVAudioPCMBuffer? {
        guard let formatDesc = CMSampleBufferGetFormatDescription(self),
              let asbd = CMAudioFormatDescriptionGetStreamBasicDescription(formatDesc)
        else { return nil }

        guard let avFormat = AVAudioFormat(streamDescription: asbd) else { return nil }

        let frameCount = CMSampleBufferGetNumSamples(self)
        guard frameCount > 0,
              let pcmBuffer = AVAudioPCMBuffer(pcmFormat: avFormat, frameCapacity: AVAudioFrameCount(frameCount))
        else { return nil }

        pcmBuffer.frameLength = AVAudioFrameCount(frameCount)

        guard let blockBuffer = CMSampleBufferGetDataBuffer(self) else { return nil }
        let length = CMBlockBufferGetDataLength(blockBuffer)
        guard let channelData = pcmBuffer.floatChannelData else { return nil }

        CMBlockBufferCopyDataBytes(blockBuffer, atOffset: 0, dataLength: length, destination: channelData[0])
        return pcmBuffer
    }
}

File: Windows/Type4Me/Audio/AudioCaptureEngine.cs
using NAudio.Wave;
using Type4Me.Services;

namespace Type4Me.Audio;

/// <summary>
/// Audio capture engine using NAudio WaveInEvent.
/// Produces 16kHz / 16-bit / mono PCM chunks (200ms = 6400 bytes).
/// </summary>
public sealed class AudioCaptureEngine : IDisposable
{
    // ── Constants ───────────────────────────────────────────

    public const int SampleRate = 16000;
    public const int Channels = 1;
    public const int BitsPerSample = 16;
    public const int ChunkDurationMs = 200;
    public const int SamplesPerChunk = 3200;
    public const int ChunkByteSize = 6400;

    // ── Events ─────────────────────────────────────────────

    /// <summary>Called on the NAudio thread with each 6400-byte PCM chunk.</summary>
    public Action<byte[]>? OnAudioChunk;

    /// <summary>Called with normalized 0..1 audio level for visualization.</summary>
    public Action<float>? OnAudioLevel;

    // ── Private state ──────────────────────────────────────

    private WaveInEvent? _waveIn;
    private readonly object _lock = new();
    private byte[] _buffer = [];
    private byte[] _accumulated = [];
    private int _levelCounter;
    private bool _isWarmedUp;

    // ── Warm-up ────────────────────────────────────────────

    /// <summary>Pre-initialize audio subsystem to reduce first-recording latency.</summary>
    public void WarmUp()
    {
        if (_isWarmedUp) return;

        Task.Run(() =>
        {
            try
            {
                var waveIn = new WaveInEvent
                {
                    WaveFormat = new WaveFormat(SampleRate, BitsPerSample, Channels),
                    BufferMilliseconds = 100,
                };
                waveIn.StartRecording();
                Thread.Sleep(300);
                waveIn.StopRecording();
                waveIn.Dispose();
                _isWarmedUp = true;
                DebugFileLogger.Log("[Audio] Warm-up complete");
            }
            catch (Exception ex)
            {
                DebugFileLogger.Log($"[Audio] Warm-up failed: {ex.Message}");
            }
        });
    }

    // ── Start / Stop ───────────────────────────────────────

    public void Start()
    {
        lock (_lock)
        {
            _buffer = [];
            _accumulated = [];
        }
        _levelCounter = 0;

        _waveIn = new WaveInEvent
        {
            WaveFormat = new WaveFormat(SampleRate, BitsPerSample, Channels),
            BufferMilliseconds = 100,
        };
        _waveIn.DataAvailable += OnDataAvailable;
        _waveIn.StartRecording();
        _isWarmedUp = true;
        DebugFileLogger.Log("[Audio] Capture started");
    }

    public void Stop()
    {
        if (_waveIn == null) return;

        _waveIn.StopRecording();
        _waveIn.DataAvailable -= OnDataAvailable;
        _waveIn.Dispose();
        _waveIn = null;

        FlushRemaining();
        _levelCounter = 0;
        DebugFileLogger.Log("[Audio] Capture stopped");
    }

    /// <summary>Returns all PCM audio accumulated since Start().</summary>
    public byte[] GetRecordedAudio()
    {
        lock (_lock)
        {
            return _accumulated.ToArray();
        }
    }

    // ── NAudio callback ────────────────────────────────────

    private void OnDataAvailable(object? sender, WaveInEventArgs e)
    {
        if (e.BytesRecorded == 0) return;

        // Audio level (~20 times/sec)
        _levelCounter++;
        if (_levelCounter % 3 == 0)
        {
            float level = CalculateLevel(e.Buffer, e.BytesRecorded);
            OnAudioLevel?.Invoke(level);
        }

        lock (_lock)
        {
            // Append to accumulated and chunk buffer
            int prevAccLen = _accumulated.Length;
            Array.Resize(ref _accumulated, prevAccLen + e.BytesRecorded);
            Buffer.BlockCopy(e.Buffer, 0, _accumulated, prevAccLen, e.BytesRecorded);

            int prevBufLen = _buffer.Length;
            Array.Resize(ref _buffer, prevBufLen + e.BytesRecorded);
            Buffer.BlockCopy(e.Buffer, 0, _buffer, prevBufLen, e.BytesRecorded);

            EmitFullChunks();
        }
    }

    /// <summary>Emit all complete 6400-byte chunks from buffer. Must be called with _lock held.</summary>
    private void EmitFullChunks()
    {
        while (_buffer.Length >= ChunkByteSize)
        {
            var chunk = new byte[ChunkByteSize];
            Buffer.BlockCopy(_buffer, 0, chunk, 0, ChunkByteSize);

            int remaining = _buffer.Length - ChunkByteSize;
            if (remaining > 0)
            {
                var newBuf = new byte[remaining];
                Buffer.BlockCopy(_buffer, ChunkByteSize, newBuf, 0, remaining);
                _buffer = newBuf;
            }
            else
            {
                _buffer = [];
            }

            OnAudioChunk?.Invoke(chunk);
        }
    }

    /// <summary>RMS → normalized 0..1 level from 16-bit PCM.</summary>
    private static float CalculateLevel(byte[] buffer, int bytesRecorded)
    {
        int sampleCount = bytesRecorded / 2;
        if (sampleCount == 0) return 0;

        double sum = 0;
        int stride = Math.Max(1, sampleCount / 256);
        int count = 0;

        for (int i = 0; i < sampleCount; i += stride)
        {
            short sample = BitConverter.ToInt16(buffer, i * 2);
            double normalized = sample / 32768.0;
            sum += normalized * normalized;
            count++;
        }

        float rms = (float)Math.Sqrt(sum / count);
        float db = 20f * MathF.Log10(MathF.Max(rms, 1e-7f));
        // Map -50dB..0dB → 0..1
        return Math.Clamp((db + 50f) / 50f, 0f, 1f);
    }

    private void FlushRemaining()
    {
        byte[] remaining;
        lock (_lock)
        {
            remaining = _buffer;
            _buffer = [];
        }

        if (remaining.Length > 0)
            OnAudioChunk?.Invoke(remaining);
    }

    public void Dispose()
    {
        Stop();
    }
}
